File: Assignment1/matrixmul.py
import numpy as np
import time

def matrix_multiplication_int(N):
    # Total system and CPU time
    start_wall_total = time.time()
    start_cpu_total = time.process_time()
    
    A = np.random.randint(0, 100, (N, N)).astype(np.int32)
    B = np.random.randint(0, 100, (N, N)).astype(np.int32)
    C = np.zeros((N, N), dtype=np.int32)

    # Meat portion (only core matrix multiplication)
    start_cpu_core = time.process_time()
    for i in range(N):
        for j in range(N):
            for k in range(N):
                C[i, j] += A[i, k] * B[k, j]
    end_cpu_core = time.process_time()

    end_wall_total = time.time()
    end_cpu_total = time.process_time()

    system_time_total = end_wall_total - start_wall_total
    cpu_time_total = end_cpu_total - start_cpu_total
    cpu_time_core = end_cpu_core - start_cpu_core

    return system_time_total, cpu_time_total, cpu_time_core

def matrix_multiplication_double(N):
    # Total system and CPU time
    start_wall_total = time.time()
    start_cpu_total = time.process_time()
    
    A = np.random.randint(0, 100, (N, N)).astype(np.float64)
    B = np.random.randint(0, 100, (N, N)).astype(np.float64)
    C = np.zeros((N, N), dtype=np.float64)

    # Meat portion (only core matrix multiplication)
    start_cpu_core = time.process_time()
    for i in range(N):
        for j in range(N):
            for k in range(N):
                C[i, j] += A[i, k] * B[k, j]
    end_cpu_core = time.process_time()

    end_wall_total = time.time()
    end_cpu_total = time.process_time()

    system_time_total = end_wall_total - start_wall_total
    cpu_time_total = end_cpu_total - start_cpu_total
    cpu_time_core = end_cpu_core - start_cpu_core

    return system_time_total, cpu_time_total, cpu_time_core

sizes = [64, 128, 256, 512, 1024]

for size in sizes:
    sys_time_int, cpu_time_total_int, cpu_time_core_int = matrix_multiplication_int(size)
    sys_time_double, cpu_time_total_double, cpu_time_core_double = matrix_multiplication_double(size)

    print(f"Matrix size {size}x{size} (Integers):")
    print(f"  Total System time: {sys_time_int:.6f} seconds")
    print(f"  Total CPU time: {cpu_time_total_int:.6f} seconds")
    print(f"  Core CPU time (meat portion): {cpu_time_core_int:.6f} seconds\n")

    print(f"Matrix size {size}x{size} (Doubles):")
    print(f"  Total System time: {sys_time_double:.6f} seconds")
    print(f"  Total CPU time: {cpu_time_total_double:.6f} seconds")
    print(f"  Core CPU time (meat portion): {cpu_time_core_double:.6f} seconds\n")

File: Assignment1/matrixmul.cpp
#include <iostream>
#include <vector>
#include <ctime>

void matrix_multiplication_int(int N) {
    struct timespec start_wall_total, end_wall_total, start_cpu_total, end_cpu_total, start_cpu_core, end_cpu_core;

    // Start total timing
    clock_gettime(CLOCK_MONOTONIC, &start_wall_total);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start_cpu_total);
    
    std::vector<std::vector<int>> A(N, std::vector<int>(N));
    std::vector<std::vector<int>> B(N, std::vector<int>(N));
    std::vector<std::vector<int>> C(N, std::vector<int>(N, 0));

    // Fill matrices with random integers
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            A[i][j] = rand() % 100, B[i][j] = rand() % 100;

    // Start core timing (meat portion)
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start_cpu_core);
    // Matrix multiplication
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            for (int k = 0; k < N; ++k)
                C[i][j] += A[i][k] * B[k][j];

    // End core timing (meat portion)
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end_cpu_core);

    // End total timing
    clock_gettime(CLOCK_MONOTONIC, &end_wall_total);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end_cpu_total);

    // Calculate elapsed time
    double wall_time_total = (end_wall_total.tv_sec - start_wall_total.tv_sec) + (end_wall_total.tv_nsec - start_wall_total.tv_nsec) / 1e9;
    double cpu_time_total = (end_cpu_total.tv_sec - start_cpu_total.tv_sec) + (end_cpu_total.tv_nsec - start_cpu_total.tv_nsec) / 1e9;
    double cpu_time_core = (end_cpu_core.tv_sec - start_cpu_core.tv_sec) + (end_cpu_core.tv_nsec - start_cpu_core.tv_nsec) / 1e9;
    
    std::cout << "Matrix size " << N << "x" << N << " (Integers):\n";
    std::cout << "  Total System time: " << wall_time_total << " seconds\n";
    std::cout << "  Total CPU time: " << cpu_time_total << " seconds\n";
    std::cout << "  Core CPU time (meat portion): " << cpu_time_core << " seconds\n";
}

void matrix_multiplication_double(int N) {
    struct timespec start_wall_total, end_wall_total, start_cpu_total, end_cpu_total, start_cpu_core, end_cpu_core;
    
    // Start total timing
    clock_gettime(CLOCK_MONOTONIC, &start_wall_total);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start_cpu_total);

    std::vector<std::vector<double>> A(N, std::vector<double>(N));
    std::vector<std::vector<double>> B(N, std::vector<double>(N));
    std::vector<std::vector<double>> C(N, std::vector<double>(N, 0));

    // Fill matrices with random doubles
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            A[i][j] = (double)(rand() % 100), B[i][j] = (double)(rand() % 100);

    // Start core timing (meat portion)
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &start_cpu_core);
    // Matrix multiplication
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            for (int k = 0; k < N; ++k)
                C[i][j] += A[i][k] * B[k][j];

    // End core timing (meat portion)
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end_cpu_core);

    // End total timing
    clock_gettime(CLOCK_MONOTONIC, &end_wall_total);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &end_cpu_total);

    // Calculate elapsed time
    double wall_time_total = (end_wall_total.tv_sec - start_wall_total.tv_sec) + (end_wall_total.tv_nsec - start_wall_total.tv_nsec) / 1e9;
    double cpu_time_total = (end_cpu_total.tv_sec - start_cpu_total.tv_sec) + (end_cpu_total.tv_nsec - start_cpu_total.tv_nsec) / 1e9;
    double cpu_time_core = (end_cpu_core.tv_sec - start_cpu_core.tv_sec) + (end_cpu_core.tv_nsec - start_cpu_core.tv_nsec) / 1e9;
    
    std::cout << "Matrix size " << N << "x" << N << " (Doubles):\n";
    std::cout << "  Total System time: " << wall_time_total << " seconds\n";
    std::cout << "  Total CPU time: " << cpu_time_total << " seconds\n";
    std::cout << "  Core CPU time (meat portion): " << cpu_time_core << " seconds\n";
}


int main() {
    srand(time(0));
    int sizes[] = {64, 128, 256, 512, 1024};
    
    for(int size : sizes) {
        matrix_multiplication_int(size);
        matrix_multiplication_double(size);
    }
    return 0;
}